File: udemy/001_validAnagram.js
function validAnagram(str1, str2) {
  // add whatever parameters you deem necessary - good luck!
  //   if (str1 === str2) {
  //     return true;
  //   }

  if (str1.length !== str2.length) {
    return false;
  }

  const tempStr1 = {};
  for (let str of str1) {
    if (!tempStr1[str]) {
      tempStr1[str] = 1;
    } else {
      tempStr1[str]++;
    }
  }

  for (let str of str2) {
    if (!tempStr1[str]) {
      return false;
    } else {
      tempStr1[str] -= 1;
    }
  }

  return true;
  //   const tempStr2 = {}
  //   for (let str of str2) {
  //     if (!tempStr2[str]) {
  //       tempStr2[str] = 1;
  //     } else {
  //       tempStr2[str]++;
  //     }
  //   }
  //   //   console.log(tempStr1, tempStr2);
  //   for (let key in tempStr1) {
  //     if (tempStr1[key] == tempStr2[key]) {
  //       return true;
  //     } else return false;
  //   }
}

console.log(validAnagram("", "")); // true
console.log(validAnagram("aaz", "zza")); // false
console.log(validAnagram("anagram", "nagaram")); // true
console.log(validAnagram("rat", "car")); // false
console.log(validAnagram("awesome", "awesom")); // false
console.log(validAnagram("amanaplanacanalpanama", "acanalmanplanpamana")); // false
console.log(validAnagram("qwerty", "qeywrt")); // true
console.log(validAnagram("texttwisttime", "timetwisttext")); // true

File: udemy/001_validAnagram.py
def valid_anagram(str1,str2):
    if len(str1) != len(str2):
        return False
    temp_str1 = {}
    temp_str2 = {}
    for str in str1:
        if str not in temp_str1:
            temp_str1[str] = 1
        else:
            temp_str1[str] += 1
    for str in str2:
        if str not in temp_str2:
            temp_str2[str] = 1
        else:
            temp_str2[str] += 1
    
    if temp_str1 == temp_str2:
        return True
    else: return False

print(valid_anagram('', ''))
print(valid_anagram('aaz', 'zza') )
print(valid_anagram('anagram', 'nagaram'))
print(valid_anagram("rat","car"))
print(valid_anagram('awesome', 'awesom'))
print(valid_anagram('amanaplanacanalpanama', 'acanalmanplanpamana'))
print(valid_anagram('qwerty', 'qeywrt'))
print(valid_anagram('texttwisttime', 'timetwisttext'))